src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,729 @@
+(** OCaml implementation of the linear two-variable equality domain.
+
+    @see <http://doi.acm.org/10.1145/2049706.2049710> A. Flexeder, M. Petter, and H. Seidl Fast Interprocedural Linear Two-Variable Equalities. *)
+
+(** Abstract states in this domain are represented by structs containing an array and an apron environment.
+    The arrays are modeled as proposed in the paper: Each variable is assigned to an index and each array element represents a linear relationship that must hold at the corresponding program point.
+    The apron environment is hereby used to organize the order of columns and variables.
+*)
+
+open Batteries
+open GoblintCil
+open Pretty
+module M = Messages
+open Apron
+open VectorMatrix
+
+module Mpqf = SharedFunctions.Mpqf
+
+module Equality = struct
+  (* (Some i, k) represents a sum of a variable with index i and the number k.
+     (None, k) represents the number k. *)
+  type t = (int option * (Z.t [@printer Z.pp_print])) [@@deriving eq, ord, hash, show]
+  let zero = (None, Z.zero)
+  let var_zero i = (Some i, Z.zero)
+  let to_int x = Z.to_int @@ snd x
+end
+
+module EqualitiesArray = struct
+  include Array
+  type t = Equality.t Array.t [@@deriving eq, ord]
+
+  let show m =
+    Array.fold_right (fun k result -> Equality.show k ^ "\n" ^ result) m ""
+
+  let hash : t -> int = Array.fold_left (fun acc a -> 31 * acc + Equality.hash a) 0
+
+  let empty () = [||]
+
+  let make_empty_array len = Array.init len (fun i -> (Some i, Z.zero))
+
+  let add_variables_to_domain m indexes = (* add new variables to domain with particular indices; translates old indices to keep consistency *)
+    (* the semantic of indexes can be retrieved from apron: https://antoinemine.github.io/Apron/doc/api/ocaml/Dim.html *)
+    if length indexes = 0 then m else
+      let next_offset_bump_list = (* an ascending list of indices, where the offset is bumped by 1 *)
+        Array.to_list indexes @ [ Array.length m ] (* terminate list with m to avoid out of bounds access *)
+      in
+      let offset_map = Array.make (Array.length m) 0 (* maps each variable to the number of variables that are added before this variable *)
+      in
+      let rec shift (offset, list) index = (* bumps offset & pops list, if/while index is heading the list *)
+        if index = List.hd list then
+          shift (offset+1, List.tl list) index
+        else (offset, list)
+      in
+      let _ =
+        Array.fold_lefti (* iterates over all indices of offset_map, overwrites content with current offset wrt. potential shift *)
+          (fun (offset, offset_bump_list) index _ ->
+             let newoffset, newlist = shift (offset, offset_bump_list) index in
+             offset_map.(index) <- newoffset;
+             (newoffset, newlist))
+          (0, next_offset_bump_list) offset_map
```

Comment:
Internal remark for refactoring by advisors after deadline: This is a fold over an array ignoring the values, the array is only used for the indices and modified by side-effects.